<a href="https://colab.research.google.com/github/etomoscow/DL-in-NLP/blob/master/hw3/task4_sentiment_cnn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment 2.4: Text classification via CNN (20 points)

In this assignment you should perform sentiment analysis of the IMDB reviews based on CNN architecture. Read carefully [Convolutional Neural Networks for Sentence Classification](https://arxiv.org/pdf/1408.5882.pdf) by Yoon Kim.

In [0]:
import numpy as np
import torch
from sklearn.metrics import f1_score, accuracy_score, precision_score, recall_score 

import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

from torchtext import datasets
from torchtext.data import Field, LabelField
from torchtext.data import BucketIterator

### Preparing Data

In [0]:
TEXT = Field(sequential=True,
             lower=True,
             batch_first=True,
             fix_length=200,
             include_lengths=True
             )
LABEL = LabelField(batch_first=True)

In [3]:
train, tst = datasets.IMDB.splits(TEXT, LABEL)
trn, vld = train.split()

downloading aclImdb_v1.tar.gz


aclImdb_v1.tar.gz: 100%|██████████| 84.1M/84.1M [00:07<00:00, 11.1MB/s]


In [4]:
%%time
TEXT.build_vocab(trn)

CPU times: user 1.36 s, sys: 37 ms, total: 1.39 s
Wall time: 1.4 s


In [0]:
LABEL.build_vocab(trn)

### Creating the Iterator (2 points)

Define an iterator here

In [0]:
train_iter, val_iter, test_iter = BucketIterator.splits(
        (trn, vld, tst),
        batch_sizes=(64, 64, 64),
        sort=False,
        sort_key= lambda x: len(x.text),
        sort_within_batch=False,
        device='cuda',
        repeat=False
)

In [0]:
batch = next(train_iter.__iter__())

### Define CNN-based text classification model (8 points)

In [0]:
class CNN(nn.Module):
    def __init__(self, V, D, kernel_sizes, stride=1, padding=0, dropout=0.5, out_channels=1):
        super(CNN, self).__init__()
        
        self.V = V
        self.D = D
        self.kernel_sizes = kernel_sizes
        self.stride = stride
        self.padding = padding
        self.dropout = dropout
        self.out_channels = out_channels
    
        self.embs = nn.Embedding(V, D, padding_idx=1)
        self.conv1 = nn.Conv2d(1, out_channels,(kernel_sizes[0], D))
        self.conv2 = nn.Conv2d(1, out_channels, (kernel_sizes[1], D))
        self.conv3 = nn.Conv2d(1, out_channels, (kernel_sizes[2], D))
        self.dp = nn.Dropout(dropout)
        self.lin = nn.Linear(len(kernel_sizes)*out_channels, 2)
    
    
    def convolution_block(self, input, conv):
        conved = conv(input)
        act = F.relu(conved.squeeze(3))
        out = F.max_pool1d(act, act.size()[2]).squeeze(2)

        return out
    def forward(self, x):
        
        x = self.embs(x).unsqueeze(1)
        out1 = self.convolution_block(x, self.conv1)
        out2 = self.convolution_block(x, self.conv2)
        out3 = self.convolution_block(x, self.conv3)

        concat = self.dp(torch.cat((out1, out2, out3), 1))
        logit = self.lin(concat)
        return logit

In [0]:
kernel_sizes = [3,4,5]
vocab_size = len(TEXT.vocab)
dropout = 0.5
dim = 300

model = CNN(vocab_size, dim, kernel_sizes, dropout)

In [10]:
model.cuda()

CNN(
  (embs): Embedding(202121, 300, padding_idx=1)
  (conv1): Conv2d(1, 1, kernel_size=(3, 300), stride=(1, 1))
  (conv2): Conv2d(1, 1, kernel_size=(4, 300), stride=(1, 1))
  (conv3): Conv2d(1, 1, kernel_size=(5, 300), stride=(1, 1))
  (dp): Dropout(p=0.5, inplace=False)
  (lin): Linear(in_features=3, out_features=2, bias=True)
)

### The training loop (3 points)

Define the optimization function and the loss functions.

In [0]:
opt = optim.Adam(model.parameters(), lr=0.05)
loss_func = nn.CrossEntropyLoss()

Think carefully about the stopping criteria. 

In [0]:
epochs = 10

In [0]:
def train_model(model, optimizer, loss_func, num_epochs, trainset, valset):
    %%time
    train_loss, valid_loss = [], []
    for epoch in range(1, num_epochs + 1):
        train_acc, val_acc = [], []
        running_loss = 0.0
        running_corrects = 0
        model.train() 
        for batch in trainset:         
        
            x = batch.text[0]
            y = batch.label
        
            opt.zero_grad()
            preds = model(x)
            loss = loss_func(preds, y)
            loss.backward()
            optimizer.step()
            running_loss += loss.item()
        
            n_correct = (torch.max(preds, 1)[1].view(y.size()).data == y.data).float().sum().detach().cpu().numpy()
            train_acc.append(100.0 * n_correct/len(batch))

        epoch_loss = running_loss / len(trn)
        train_loss.append(epoch_loss)

        val_loss = 0.0
        model.eval()
        correct = 0
        total = 0 
        for batch in val_iter:
            with torch.no_grad():
                x = batch.text[0]
                y = batch.label
        
                preds = model(x)
                loss = loss_func(preds, y)
                val_loss += loss.item()

                n_correct = (torch.max(preds, 1)[1].view(y.size()).data == y.data).float().sum().detach().cpu().numpy()
                val_acc.append(100.0 * n_correct/len(batch))

        val_loss /= len(vld)
        valid_loss.append(val_loss)
        print('Epoch: {}, Training Loss: {}, Training Accuracy: {}, Validation Loss: {}, Validation Accuracy: {}'.format(epoch, round(epoch_loss, 5), round(np.mean(train_acc), 2), round(val_loss, 5), round(np.mean(val_acc), 2)))

In [14]:
train_model(model, opt, loss_func, epochs, train_iter, val_iter)

CPU times: user 2 µs, sys: 0 ns, total: 2 µs
Wall time: 5.01 µs
Epoch: 1, Training Loss: 0.01241, Training Accuracy: 49.87, Validation Loss: 0.01186, Validation Accuracy: 51.25
Epoch: 2, Training Loss: 0.01275, Training Accuracy: 49.73, Validation Loss: 0.01092, Validation Accuracy: 49.38
Epoch: 3, Training Loss: 0.01259, Training Accuracy: 50.1, Validation Loss: 0.01286, Validation Accuracy: 50.06
Epoch: 4, Training Loss: 0.0138, Training Accuracy: 50.81, Validation Loss: 0.01233, Validation Accuracy: 50.4
Epoch: 5, Training Loss: 0.01426, Training Accuracy: 51.14, Validation Loss: 0.01286, Validation Accuracy: 50.63
Epoch: 6, Training Loss: 0.01549, Training Accuracy: 51.58, Validation Loss: 0.01431, Validation Accuracy: 51.94
Epoch: 7, Training Loss: 0.01504, Training Accuracy: 51.21, Validation Loss: 0.01112, Validation Accuracy: 51.48
Epoch: 8, Training Loss: 0.01529, Training Accuracy: 51.6, Validation Loss: 0.01521, Validation Accuracy: 49.39
Epoch: 9, Training Loss: 0.01795, Tr

### Calculate performance of the trained model (2 points)

In [0]:
preds = model(batch.text[0])

In [20]:
torch.max(preds, 1)[1]

tensor([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1,
        0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], device='cuda:0')

In [21]:
def evaluate_model(model, testset):
    tn = 0
    tp = 0
    fp = 0
    fn = 0
    total = 0
    model.eval()
    accuracies, f1_scores, precisions, recalls = [], [], [], []
    for batch in testset:
        x = batch.text[0]
        y = batch.label
        preds = model(x)
        preds = torch.max(preds, 1)[1].view(y.size()).data.detach().cpu().numpy()
        target = y.data.detach().cpu().numpy()
        total += len(target)
        tp += (preds * target).sum()
        tn += ((1 - preds) * (1 - target)).sum()
        fp += (preds*(1 - target)).sum()
        fn += ((1-preds)*target).sum()

    recall = tp /float(tp+fn)
    precision = tp/float(tp+fp)
    f1 = (2*tp)/float(2*tp+fp+fn)
    accuracy = (tp+tn) / float(total)

    print('Average classification scores:\nAccuracy:{}, \nPrecision:{},\nRecall:{},\nF1-score:{}'.format(round(accuracy,3),
                                                                                                     round(f1,3),
                                                                                                     round(precision,3),
                                                                                                     round(recall,3)))

ERROR! Session/line number was not unique in database. History logging moved to new session 59


In [22]:
evaluate_model(model, test_iter)

Average classification scores:
Accuracy:0.513, 
Precision:0.172,
Recall:0.573,
F1-score:0.101


Write down the calculated performance

### Accuracy:0.513
### Precision:0.172
### Recall: 0.573
### F1: 0.101

### Experiments (5 points)

Experiment with the model and achieve better results. Implement and describe your experiments in details, mention what was helpful.

### 1. Bad results on the baseline. Let's tune the model. Actions: lower the embedding dimension, lower the learning rate, increase the number of epochs.Results: Accuracy:0.693, Precision:0.701, Recall:0.683, F1-score:0.721

In [0]:
kernel_sizes = [3,4,5]
vocab_size = len(TEXT.vocab)
dropout = 0.5
dim = 256

model = CNN(vocab_size, dim, kernel_sizes, dropout)
model.cuda()
opt = optim.Adam(model.parameters(), lr=0.0002)
loss_func = nn.CrossEntropyLoss()

In [24]:
train_model(model, opt, loss_func, 30, train_iter, val_iter)

CPU times: user 2 µs, sys: 1e+03 ns, total: 3 µs
Wall time: 7.15 µs
Epoch: 1, Training Loss: 0.01364, Training Accuracy: 50.99, Validation Loss: 0.01085, Validation Accuracy: 53.15
Epoch: 2, Training Loss: 0.01129, Training Accuracy: 52.07, Validation Loss: 0.01074, Validation Accuracy: 57.3
Epoch: 3, Training Loss: 0.01088, Training Accuracy: 54.24, Validation Loss: 0.01059, Validation Accuracy: 60.3
Epoch: 4, Training Loss: 0.01071, Training Accuracy: 56.22, Validation Loss: 0.01052, Validation Accuracy: 60.93
Epoch: 5, Training Loss: 0.01059, Training Accuracy: 57.17, Validation Loss: 0.01038, Validation Accuracy: 62.75
Epoch: 6, Training Loss: 0.01044, Training Accuracy: 58.62, Validation Loss: 0.01023, Validation Accuracy: 64.35
Epoch: 7, Training Loss: 0.01033, Training Accuracy: 59.39, Validation Loss: 0.0102, Validation Accuracy: 63.97
Epoch: 8, Training Loss: 0.01019, Training Accuracy: 60.51, Validation Loss: 0.01005, Validation Accuracy: 65.4
Epoch: 9, Training Loss: 0.01012

In [25]:
evaluate_model(model, test_iter)

Average classification scores:
Accuracy:0.693, 
Precision:0.701,
Recall:0.683,
F1-score:0.721
